src/accountsdb/snapshots.zig

```diff
@@ -2215,44 +2214,72 @@ pub const SnapshotFiles = struct {
+            const filename = dir_entry.name;
 
-        // find the snapshots
-        var maybe_latest_full_snapshot: ?FullSnapshotFileInfo = null;
-        var count: usize = 0;
-        for (files) |filename| {
-            const snapshot = FullSnapshotFileInfo.parseFileNameTarZst(filename) catch continue;
-            if (count == 0 or snapshot.slot > maybe_latest_full_snapshot.?.slot) {
-                maybe_latest_full_snapshot = snapshot;
+            if (IncrementalSnapshotFileInfo.parseFileNameTarZst(filename)) |inc_snap| {
+                if (maybe_latest_full) |latest_full| {
+                    if (inc_snap.slot < latest_full.slot) continue;
+                    if (inc_snap.base_slot < latest_full.slot) continue;
+                }
+                try incremental_snaps.append(allocator, inc_snap);
+                continue;
+            } else |_| {}
+
+            const current = FullSnapshotFileInfo.parseFileNameTarZst(filename) catch continue;
+            const prev = maybe_latest_full orelse {
+                maybe_latest_full = current;
+                continue;
+            };
+            if (prev.slot < current.slot) {
+                maybe_latest_full = current;
+                continue;
+            }
+            if (prev.slot == current.slot) {
+                // TODO:
+                std.debug.panic("TODO: report this error gracefully in some way ({s} vs {s})", .{
+                    prev.snapshotArchiveName().constSlice(),
+                    current.snapshotArchiveName().constSlice(),
+                });
             }
-            count += 1;
         }
-        const latest_full_snapshot = maybe_latest_full_snapshot orelse
-            return error.NoFullSnapshotFileInfoFound;
-
-        count = 0;
-        var maybe_latest_incremental_snapshot: ?IncrementalSnapshotFileInfo = null;
-        for (files) |filename| {
-            const snapshot = IncrementalSnapshotFileInfo.parseFileNameTarZst(filename) catch continue;
-            // need to match the base slot
-            if (snapshot.base_slot == latest_full_snapshot.slot and (count == 0 or
-                // this unwrap is safe because count > 0
-                snapshot.slot > maybe_latest_incremental_snapshot.?.slot))
-            {
-                maybe_latest_incremental_snapshot = snapshot;
+        const latest_full = maybe_latest_full orelse return error.NoFullSnapshotFileInfoFound;
+
+        var maybe_latest_incremental: ?IncrementalSnapshotFileInfo = null;
+        for (incremental_snaps.items) |current| {
+            if (current.base_slot != latest_full.slot) continue;
+            const prev = maybe_latest_incremental orelse {
+                maybe_latest_incremental = current;
+                continue;
+            };
+            if (prev.slot < current.slot) {
+                maybe_latest_incremental = current;
+                continue;
+            }
+            if (prev.slot == current.slot) {
```

Comment:
why would this be an error?

Reply:
If they're the same slot, it means they're different hashes, since they're different file names. If there are two snapshots with the same snapshot range, but different hashes, it's not possible to unambiguously select one; rather call it out than just defer to some arbitrary behavior like selecting the first one that the filesystem reports.

Reply:
good point - can we add a comment explaining this above the if statement